doc/developer/design/20320328_persist_columnar.md

```diff
@@ -0,0 +1,917 @@
+instead we convert to the session timezone when loaded. This is how both
+Postgres works.
+
+</td>
+
+<tr>
+<td> Interval </td>
+<td>
+
+```rust
+struct Interval {
+  months: i32,
+  days: i32,
+  micros: i64,
+}
+```
+
+</td>
+<td>
+
+`FixedSizeBinary[16]`
+
+</td>
+<td>
+
+Represented by encoding the `months`, `days`, and `micros` fields encoded as
+big endian.
+
+> **Alternative:** The smallest possible representation for interval would be
+11 bytes, or 12 if we don't want to do bit swizzling. But other than space
+savings I don't believe there is a benefit to this approach. In fact it would
+incur some computational overhead to encode and there are no benefits from a
+SIMD perspective either.
+
+> **Alternative:** We could represent `Interval`s in a `StructArray` but we
+don't expose the internal details of `Interval` so this wouldn't aid in
+filtering or pushdown. The only benefit of structuring an interval would be for
+space reduction in dictionary encoding, but this is offset by the overhead of
+the struct itself.
+
+</td>
+
+<tr>
+<td> Jsonb </td>
+<td>
+
+```rust
+// JsonbRef<'a>(Datum<'a>)
+enum Value {
+  Null,
+  Boolean(bool),
+  Number(f64),
+  String(String),
+  Array(Vec<Value>),
+  Map(BTree<String, Value>),
+}
+```
+
+</td>
+<td>
+
+`BinaryArray`
+
+</td>
+<td>
+
+Serialize JSON with the existing protobuf types, i.e. ProtoDatum, and store
+this binary blob.
+
+> **Structured Data:** An option is to structure the JSON data using an Arrow
+Union type. What is nice about this approach is it would allow us to do some
+form of projection pushdown on the JSON data. The main issue though is Arrow
+does not really support recursive data types. In fact, it is impossible to
+statically define the above `Value` enum in Arrow. The only option is to
+dynamically generate a DataType/schema given a column of values, see [1] for an
+example of this approach. I don't believe dynamically generating the schema is
+a good option because it is relatively complex, and we would end up with
+arbitrarily deep schemas based on user provided data. The arbitrarily deep
+schemas particularly concerns me because it would have unpredictable
+performance.
+
+> **Alternative:** An alternative to fully structing the data is structuing it
+with a depth limit. For example, structuring up-to X levels deep, and then
+binary encoding the rest. This gets us predictable performance with the ability
+to do limited pushdown, at the cost of code complexity. This is probably the
+best approach in the long term, but in my opinion the additional technical
+complexity makes it out-of-scope for the initial implementation.
+
+> **Alternative:** Instead of serializing the JSON data with protobuf, we could
+use a different serialization format like [BSON](https://bsonspec.org/). This
+approach is nice because it gets us a path to entirely eliminating `ProtoDatum`
+(🔥) but I am slightly leaning away from this given Protobuf is already used so
+heavily in our codebase.
```

Comment:
One thing to be careful of with BSON: whether it supports our numbers. Our JSON datums use `numeric` for number data, which may be wider than what BSON supports.

Reply:
Good callout! I added this as a note